plugins/pipelines/app/models/concerns/samson_pipelines/stage_concern.rb

```diff
@@ -10,6 +12,27 @@ def next_stages
 
   protected
 
+  # Needs to find all the possible stages in case this is a pipeline of pipelines as each subsequent stage
+  # could have valid next_stage_ids
+  def all_next_stages
+    existing_stages = next_stages.collect(&:id).to_set
+    return [] if existing_stages.empty?
+    loop do
+      last_count = existing_stages.count
+      existing_stages.each do |stage_id|
+        existing_stages += Stage.find(stage_id).next_stage_ids.map(&:to_i).to_set
+      end
+      break if last_count == existing_stages.count
+    end
+    # if we find ourselves with ourselves in the list, we have a loop.
+    # this shouldn't happen, but this will prevent it
+    if existing_stages.include?(id)
```

Comment:
me too! :D

On Fri, Apr 15, 2016 at 1:25 PM, ian waters <notifications@github.com>
wrote:

> In plugins/pipelines/app/models/concerns/samson_pipelines/stage_concern.rb
> <https://github.com/zendesk/samson/pull/780#discussion_r59935772>:
>
> > @@ -10,6 +12,27 @@ def next_stages
> >
> >    protected
> >
> > +  # Needs to find all the possible stages in case this is a pipeline of pipelines as each subsequent stage
> > +  # could have valid next_stage_ids
> > +  def all_next_stages
> > +    existing_stages = next_stages.collect(&:id).to_set
> > +    return [] if existing_stages.empty?
> > +    loop do
> > +      last_count = existing_stages.count
> > +      existing_stages.each do |stage_id|
> > +        existing_stages += Stage.find(stage_id).next_stage_ids.map(&:to_i).to_set
> > +      end
> > +      break if last_count == existing_stages.count
> > +    end
> > +    # if we find ourselves with ourselves in the list, we have a loop.
> > +    # this shouldn't happen, but this will prevent it
> > +    if existing_stages.include?(id)
>
> Honestly I'd like to just to back to the recursive code that was there and
> rely on the model validation to prevent a loop (as it was before).
>
> thoughts?
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zendesk/samson/pull/780/files/d0fdde5ea30d3bb6864caed74a69a03a3b21d6a3#r59935772>
>